Allow customisation of JWT expiration duration

albertodebortoli opened the pull request:
This PR allows customizing the expiration duration of the token in `APIConfiguration`. This is to account for possible clock skews.

From the documentation:

> exp - Expiration Time | The token’s expiration time in Unix epoch time. Tokens that expire more than 20 minutes into the future are not valid except for resources listed in Determine the Appropriate Token Lifetime.

Source: [Generating Tokens for API Requests](https://developer.apple.com/documentation/appstoreconnectapi/generating_tokens_for_api_requests)

Using the default value of 20 minutes might be too strict in cases where there is a slight clock difference between the client and the server (where the clock of the client machine is slightly ahead of the clock of the server), resulting in the server refusing the request with the following API error:

```
Error: Request https://api.appstoreconnect.apple.com/v1/users?<REDACTED> failed with status code 401. Related response error(s):

The request failed with response code 401 NOT_AUTHORIZED

Authentication credentials are missing or invalid.. Provide a properly configured and signed bearer token, and make sure that it has not expired. Learn more about Generating Tokens for API Requests https://developer.apple.com/go/?id=api-generating-tokens).
```

The error above can easily be reproduce by fabricating a skewed date in `JWT.swift` while leaving the default expiration duration of 20 minutes:

```
static let defaultDateProvider: DateProvider = {
    Date().addingTimeInterval(60)
}
```

These changes maintain backward compatibility.

AvdLee commented on README.md:
````diff
@@ -46,13 +46,13 @@ Copy the contents and remove the whitelines, `-----BEGIN PRIVATE KEY-----` and `
 
 ```swift
-let configuration = APIConfiguration(issuerID: "<YOUR ISSUER ID>", privateKeyID: "<YOUR PRIVATE KEY ID>", privateKey: "<YOUR PRIVATE KEY>")
+let configuration = APIConfiguration(issuerID: "<YOUR ISSUER ID>", privateKeyID: "<YOUR PRIVATE KEY ID>", privateKey: "<YOUR PRIVATE KEY>", expirationDuration: "<OPTIONAL EXPIRE DURATION>")
````
Since it's optional and likely not used by everyone, I think it's better to keep the minimum configuration sample in the readme. Thoughts?

```suggestion
let configuration = APIConfiguration(issuerID: "<YOUR ISSUER ID>", privateKeyID: "<YOUR PRIVATE KEY ID>", privateKey: "<YOUR PRIVATE KEY>")
```

AvdLee commented on README.md:
````diff
@@ -46,13 +46,13 @@ Copy the contents and remove the whitelines, `-----BEGIN PRIVATE KEY-----` and `
 
 ```swift
-let configuration = APIConfiguration(issuerID: "<YOUR ISSUER ID>", privateKeyID: "<YOUR PRIVATE KEY ID>", privateKeyURL: URL(fileURLWithPath: "~/AuthKey_<YOUR PRIVATE KEY ID>.p8"))
+let configuration = APIConfiguration(issuerID: "<YOUR ISSUER ID>", privateKeyID: "<YOUR PRIVATE KEY ID>", privateKeyURL: URL(fileURLWithPath: "~/AuthKey_<YOUR PRIVATE KEY ID>.p8"), expirationDuration: "<OPTIONAL EXPIRE DURATION>")
````
Removing it here too.

```suggestion
let configuration = APIConfiguration(issuerID: "<YOUR ISSUER ID>", privateKeyID: "<YOUR PRIVATE KEY ID>", privateKeyURL: URL(fileURLWithPath: "~/AuthKey_<YOUR PRIVATE KEY ID>.p8"))
```

Even better: how about adding a separate code example in the readme with a bit of explanation on when to use the expiration duration?

AvdLee commented on Sources/APIProvider.swift:
```diff
@@ -29,15 +29,19 @@ public struct APIConfiguration {
-    public init(issuerID: String, privateKeyID: String, privateKey: String) throws {
+    ///   - privateKeyID: Your private key ID from App Store Connect (Ex: 2X9R4HXF34)
+    ///   - privateKey: Your private key stripped out of the -----BEGIN PRIVATE KEY----- and -----END PRIVATE KEY----- lines.
+    ///   - expirationDuration: The token's expiration duration in seconds. Tokens that expire more than 20 minutes in the future are not valid, so set it to a max of 20 minutes.
```
```suggestion
    ///   - expirationDuration: The token's expiration duration in seconds. Tokens that expire more than 20 minutes in the future are not valid, so set it to a max of 20 minutes. Defaults to 20 minutes.
```

AvdLee commented on Sources/APIProvider.swift:
```diff
@@ -54,7 +59,8 @@ public struct APIConfiguration {
     ///   - privateKeyID: Your private key ID from App Store Connect (Ex: 2X9R4HXF34). Will be inferred from `privateKeyURL` if nil.
     ///   - privateKeyURL: A file URL that references the path to your private key file.
-    public init(issuerID: String, privateKeyID: String? = nil, privateKeyURL: URL) throws {
+    ///   - expirationDuration: The token's expiration duration in seconds. Tokens that expire more than 20 minutes in the future are not valid, so set it to a max of 20 minutes.
```
```suggestion
    ///   - expirationDuration: The token's expiration duration in seconds. Tokens that expire more than 20 minutes in the future are not valid, so set it to a max of 20 minutes. Defaults to 20 minutes.
```

AvdLee reviewed (changes requested):
Thanks a lot for your contribution, this makes sense! Do you think you're able to write a test to cover the custom expiration duration? I've also added a few suggestions in the comments. Let me know what you think!

albertodebortoli commented:
Hi @AvdLee thanks a lot for the review!
I believe I’ve addressed all your suggestions and added a test that I should have included in my initial attempt 😄.
Please let me know what you think. I’m happy to rephrase the README if you have any additional observations.

albertodebortoli commented:
Hi @AvdLee any chance you can take another look at this?

albertodebortoli commented:
@AvdLee 👋

albertodebortoli commented:
@AvdLee Hello hello!

AvdLee commented:
@albertodebortoli Thanks for implementing the feedback! This slipped through my overview, sorry for the late response!

albertodebortoli commented:
OMG how is it possible I didn't spot the `.orig` file.
